include/audio/aseq.h

```diff
@@ -0,0 +1,1979 @@
+    /* Helper for writing command ids. Will write the command id appropriate for the current section, optionally
+     * packing in low-order bits if a command uses them for an argument.                                         */
+    _check_arg_bitwidth_u \lo_bits, \n_lo_bits
+
+    .if ASEQ_MODE == ASEQ_MODE_SEQUENCE && \seq_id != -1
+        _check_arg_bitwidth_u \seq_id, 8
+        .byte (\seq_id & 0xFF) | (\lo_bits)
+    .elseif ASEQ_MODE == ASEQ_MODE_CHANNEL && \chan_id != -1
+        _check_arg_bitwidth_u \chan_id, 8
+        .byte (\chan_id & 0xFF) | (\lo_bits)
+    .elseif ASEQ_MODE == ASEQ_MODE_LAYER && \layer_id != -1
+        _check_arg_bitwidth_u \layer_id, 8
+        .byte (\layer_id & 0xFF) | (\lo_bits)
+    .elseif ASEQ_MODE == ASEQ_MODE_TABLE && \tbl_id != -1
+        _check_arg_bitwidth_u \tbl_id, 8
+        .byte (\tbl_id & 0xFF) | (\lo_bits)
+    .elseif ASEQ_MODE == ASEQ_MODE_ARRAY && \arr_id != -1
+        _check_arg_bitwidth_u \arr_id, 8
+        .byte (\arr_id & 0xFF) | (\lo_bits)
+    .elseif ASEQ_MODE == ASEQ_MODE_FILTER && \filt_id != -1
+        _check_arg_bitwidth_u \filt_id, 8
+        .byte (\filt_id & 0xFF) | (\lo_bits)
+    .elseif ASEQ_MODE == ASEQ_MODE_ENVELOPE && \env_id != -1
+        _check_arg_bitwidth_u \env_id, 8
+        .byte (\env_id & 0xFF) | (\lo_bits)
+    .elseif ASEQ_MODE == ASEQ_MODE_BUFFER && \buf_id != -1
+        _check_arg_bitwidth_u \buf_id, 8
+        .byte (\buf_id & 0xFF) | (\lo_bits)
+    .else
+        _section_invalid \cmd_name
+    .endif
+.endm
+
+.macro .ptr value
+    _wr_lbl \value
+.endm
+
+.macro .ptr_raw value
+    _wr16 \value
+.endm
+
+
+
+/**
+ *  Instructions
+ */
+
+/**
+ * end
+ *
+ *  Marks the end of an executable section.
+ */
+.macro end
+    _wr_cmd_id  end, 0xFF,0xFF,0xFF,,,,,, 0, 0
+.endm
+
+/**
+ * delay1
+ *
+ *  Delays for one tick.
+ */
+.macro delay1
+    _wr_cmd_id  delay1, 0xFE,0xFE,,,,,,, 0, 0
+.endm
+
+/**
+ * delay <delay:var>
+ *
+ *  Delays for `delay` ticks.
+ */
+.macro delay delay
+    _wr_cmd_id  delay, 0xFD,0xFD,,,,,,, 0, 0
+    _var        \delay
+.endm
+
+/**
+ * call <label:lbl>
+ *
+ *  Unconditionally enters a subroutine at `label`. Execution will resume following this instruction once the
+ *  subroutine encounters an `end` instruction.
+ */
+.macro call label
+    _wr_cmd_id  call, 0xFC,0xFC,0xFC,,,,,, 0, 0
+    _wr_lbl     \label
+.endm
+
+/**
+ * jump <label:lbl>
+ *
+ *  Branches to `label` unconditionally.
+ */
+.macro jump label
+    _wr_cmd_id  jump, 0xFB,0xFB,0xFB,,,,,, 0, 0
+    _wr_lbl     \label
+.endm
+
+/**
+ * beqz <label:lbl>
+ *
+ *  Branches to `label` if TR == 0.
+ */
+.macro beqz label
+    _wr_cmd_id  beqz, 0xFA,0xFA,0xFA,,,,,, 0, 0
+    _wr_lbl     \label
+.endm
+
+/**
+ * bltz <label:lbl>
+ *
+ *  Branches to `label` if TR < 0.
+ */
+.macro bltz label
+    _wr_cmd_id  beqz, 0xF9,0xF9,0xF9,,,,,, 0, 0
+    _wr_lbl     \label
+.endm
+
+/**
+ * loop <num:u8>
+ *
+ *  Begin a loop for `num` iterations.
```

Comment:
In general these descriptions are a bit inconsistent with e.g. "Begin" vs "Begins" and whether they have a trailing period